runners/core-java/src/main/java/org/apache/beam/sdk/util/DoFnRunnerBase.java

```diff
@@ -352,10 +349,8 @@ private String generateInternalAggregatorName(String userName) {
     @Override
     protected <AggInputT, AggOutputT> Aggregator<AggInputT, AggOutputT> createAggregatorInternal(
         String name, CombineFn<AggInputT, ?, AggOutputT> combiner) {
-      checkNotNull(combiner,
-          "Combiner passed to createAggregator cannot be null");
-      return new CounterAggregator<>(generateInternalAggregatorName(name),
-          combiner, addCounterMutator);
+      checkNotNull(combiner, "Combiner passed to createAggregator cannot be null");
+      return aggregatorFactory.createAggregator(name, combiner);
```

Comment:
This should be `aggregatorFactory.createAggregator(generateInternalAggregatorName(name), combiner);` to ensure we associate each aggregator with the current step context.